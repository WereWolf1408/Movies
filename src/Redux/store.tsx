import { configureStore } from '@reduxjs/toolkit';
import { useDispatch } from 'react-redux';
import movies from './reducers/moviesReducer';
import modalWindows from './reducers/modalWindowsReducer';
import additionalStore from './reducers/additionalReducer';

export const store = configureStore({
  reducer: {
    movies,
    additionalStore,
    modalWindows,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export const AppDispatch = () => useDispatch<typeof store.dispatch>();
