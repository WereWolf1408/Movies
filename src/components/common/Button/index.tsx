import React from 'react';
const classnames = require('classnames');

import './style.less';

const CLASSES = {
  NETFLIX_BUTTON: 'netflix-app__button',
};

interface ButtonProps {
  (props: { text: string; classes?: string; clickHandler?: () => void }): JSX.Element;
}

const Button: ButtonProps = ({ text, classes, clickHandler = null }) => {
  return (
    <div
      onClick={clickHandler}
      className={classnames(CLASSES.NETFLIX_BUTTON, classes)}
    >
      {text}
    </div>
  );
};

export default Button;