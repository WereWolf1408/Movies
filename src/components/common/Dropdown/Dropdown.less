@import '@styles/colors.less';

.netflix-app__dropdown {
  position: relative;
  display: flex;
  flex-direction: column;
  font-family: Montserrat;
  font-style: normal;
  font-weight: 500;
  font-size: 1.6rem;
  line-height: 2rem;
  letter-spacing: .08rem;
  text-transform: uppercase;
  cursor: pointer;

  &.netflix-app__dropdown-field {
    .netflix-app__input {
      flex-grow: 1;
    }

    .netflix-app__dropdown-options {
      top: 7rem;

      span {
        margin: .5rem 0 1rem 2rem;
      }
    }

    .active {
      span {
        color: @color-carnation;
      }
    }
  }

  &-options {
    position: absolute;
    display: flex;
    flex-direction: column;
    width: 100%;
    top: 3rem;
    color: @color-white;
    background-color: @color-mine-shaft;

    span {
      margin: .5rem 0 1rem 0;

      &:hover {
        color: @color-carnation;
      }

      &.active {
        color: @color-carnation;
      }
    }
  }

  &-checkbox-label {
    display: flex;
    align-items: center;
  }

  &-checkbox {
    height: 1.6rem;
    width: 1.6rem;
    margin: 0 1rem;
  }
}
