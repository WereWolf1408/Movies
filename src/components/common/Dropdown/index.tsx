import React, { useState } from 'react';
import { DropdownCallbackProps } from '@utils/interfaces';

import './style.less';
import classNames from 'classnames';

const CLASSES = {
  NETFLIX_APP_DROPDOWN: 'netflix-app__dropdown',
  NETFLIX_APP_DROPDOWN_OPTIONS: 'netflix-app__dropdown-options',
  NETFLIX_APP_DROPDOWN_FIELD: 'netflix-app__dropdown-field',
  NETFLIX_APP_DROPDOWN_CHECKBOX: 'netflix-app__dropdown-checkbox',
  NETFLIX_APP_DROPDOWN_CHECKBOX_LABEL: 'netflix-app__dropdown-checkbox-label',
};

interface DropdownProps {
  (props: {
    options: Array<string>;
    callback?: DropdownCallbackProps;
    dropdownType: 'simple' | 'multiline';
  }): JSX.Element;
}

const Dropdown: DropdownProps = ({ callback, options, dropdownType }) => {
  const [selected, setSelected] = useState<string>(options[0]);
  const [isOpen, setIsOpen] = useState(false);
  const [optionsHolder, setOptionsHolder] = useState([options[0]]);

  const clickHandler = (option: string) => {
    setSelected(option);
    setIsOpen(false);
    callback && callback(option);
  };

  const multilineClickHandler = (
    option: string,
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    const { checked } = event.currentTarget;
    const modifyArr = (): Array<string> => {
      if (checked) {
        optionsHolder.push(option);
        return [...optionsHolder];
      }
      const filteredArr = optionsHolder.filter((value) => value !== option);
      return filteredArr;
    };
    const changedArr = modifyArr();
    setOptionsHolder(changedArr);
    setSelected(changedArr.toString());
    callback && callback(changedArr);
  };

  const selectedOptionClickHandler = () => {
    setIsOpen((prevVal) => !prevVal);
  };

  const renderDropdownOutputElement = () => {
    if (dropdownType === 'simple') {
      return <span onClick={selectedOptionClickHandler}>{selected}</span>;
    }
    return (
      <input
        className={'netflix-app__input'}
        readOnly={true}
        onClick={selectedOptionClickHandler}
        value={selected}
      />
    );
  };

  return (
    <div
      className={classNames(CLASSES.NETFLIX_APP_DROPDOWN, {
        [CLASSES.NETFLIX_APP_DROPDOWN_FIELD]: dropdownType === 'multiline',
      })}
    >
      {renderDropdownOutputElement()}
      {isOpen && (
        <div className={CLASSES.NETFLIX_APP_DROPDOWN_OPTIONS}>
          {options.map((item, index) => {
            return dropdownType === 'simple' ? (
              <span
                key={index}
                className={classNames({ ['active']: item === selected })}
                onClick={() => clickHandler(item)}
              >
                {item}
              </span>
            ) : (
              <div
                key={index}
                className={classNames({ ['active']: optionsHolder.includes(item) })}
              >
                <label
                  htmlFor=""
                  className={CLASSES.NETFLIX_APP_DROPDOWN_CHECKBOX_LABEL}
                >
                  <input
                    className={CLASSES.NETFLIX_APP_DROPDOWN_CHECKBOX}
                    type="checkbox"
                    value={item}
                    checked={optionsHolder.includes(item)}
                    onChange={(event) => multilineClickHandler(item, event)}
                  />
                  {item}
                </label>
              </div>
            );
          })}
        </div>
      )}
    </div>
  );
};

export default Dropdown;
