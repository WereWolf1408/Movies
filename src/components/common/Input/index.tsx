import React from 'react';

const classnames = require('classnames');

import './style.less';

const CLASSES = {
  NETFLIX_INPUT: 'netflix-app__input',
};

interface InputProps {
  (props: {
    changeHandler: (event: React.FormEvent<HTMLInputElement>) => void;
    placeHolder?: string;
    classes?: string;
    value?: string;
  }): JSX.Element;
}

const Input: InputProps = ({
  placeHolder = 'What do you want to watch?',
  classes,
  changeHandler,
  value,
}) => {
  return (
    <input
      className={classnames(CLASSES.NETFLIX_INPUT, classes)}
      placeholder={placeHolder}
      onChange={changeHandler}
      value={value}
    />
  );
};

export default Input;