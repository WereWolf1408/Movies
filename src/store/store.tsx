import { configureStore } from '@reduxjs/toolkit';
import { useDispatch } from 'react-redux';
import { rootReducer } from './rootReducer';
import { additionalReducer } from './additionalReducer';

export const store = configureStore({
  reducer: {
    mainStore: rootReducer,
    additionalStore: additionalReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export const AppDispatch = () => useDispatch<typeof store.dispatch>();
