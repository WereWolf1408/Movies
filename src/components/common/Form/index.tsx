import React, { Dispatch, SetStateAction, useState } from 'react';
import {
  DateInput,
  InputWithLabel,
  DropdownInput,
  TextAreaWithLabel,
} from '../SmartInputs';
import Button from '../Button';
import { formDataReceived } from '@utils/interfaces';

import './style.less';

const CLASSES = {
  NETFLIX_APP_MODAL_WINDOW_FORM: 'netflix-app__modal-window-form',
  NETFLIX_APP_MODAL_WINDOW_FORM_INPUT_SECTION:
    'netflix-app__modal-window-form-input-section',
  NETFLIX_APP_MODAL_WINDOW_FORM_INPUT_BLOCK:
    'netflix-app__modal-window-form-input-block',
  NETFLIX_APP__INPUT_BLOCK: 'netflix-app__input-block',
  NETFLIX_APP_MODAL_WINDOW_FORM_BUTTONS_SECTION:
    'netflix-app__modal-window-form-buttons-section',
};

type FormEvent = React.FormEvent<HTMLInputElement>;

type SetFormFieldProps = (
  event: React.FormEvent<HTMLInputElement>,
  callback: Dispatch<SetStateAction<string>>
) => void;

interface FormProps {
  (props: {
    resetButtonClickHandler: () => void;
    submitClickHandler: (opt: formDataReceived) => void;
  }): JSX.Element;
}

//Consider usage of useForm hook in conjunction with yup() or any other joi.
//JOI - in simple vords - validation scheme

const Form: FormProps = ({ resetButtonClickHandler, submitClickHandler }) => {
  const [title, setTitle] = useState('');
  const [movieURL, setMovieURL] = useState('');
  const [genre, setGenre] = useState(['']);
  const [releaseDate, setReleaseDate] = useState('');
  const [rating, setRating] = useState<number>(0);
  const [runTime, setRunTime] = useState('');

  const setFormField: SetFormFieldProps = (event, callback) => {
    const { value } = event.currentTarget;
    callback(value);
  };

  const titleChangeHandler = (event: FormEvent) => {
    setFormField(event, setTitle);
  };

  const movieURLChangeHandler = (event: FormEvent) => {
    setFormField(event, setMovieURL);
  };

  const genreChangeHandler = (value: Array<string>) => {
    setGenre(value);
  };

  const releaseDateChangeHandler = (event: FormEvent) => {
    setFormField(event, setReleaseDate);
  };

  const ratingChangeHandler = (event: FormEvent) => {
    const { value } = event.currentTarget;
    setRating(parseInt(value));
  };

  const runTimeChangeHandler = (event: FormEvent) => {
    setFormField(event, setRunTime);
  };

  const overviewChangeHandler = (event: React.FormEvent<HTMLTextAreaElement>) => {};

  const subminButtonClickHandler = () => {
    submitClickHandler({
      title,
      movieURL,
      genre,
      releaseDate,
      rating,
      runTime,
    });
  };

  return (
    <form className={CLASSES.NETFLIX_APP_MODAL_WINDOW_FORM}>
      <div className={CLASSES.NETFLIX_APP_MODAL_WINDOW_FORM_INPUT_SECTION}>
        <InputWithLabel
          changeHandler={titleChangeHandler}
          placeholder={'1123'}
          labelText={'Title'}
          value={title}
          classes={CLASSES.NETFLIX_APP__INPUT_BLOCK}
        />
        <InputWithLabel
          changeHandler={movieURLChangeHandler}
          placeholder={'movie url'}
          labelText={'movie url'}
          value={movieURL}
          classes={CLASSES.NETFLIX_APP__INPUT_BLOCK}
        />
        <DropdownInput
          labelText={'genre'}
          options={['comedy', 'horor', 'crime']}
          changeHandler={genreChangeHandler}
          classes={CLASSES.NETFLIX_APP__INPUT_BLOCK}
        />
        <DateInput
          labelText={'release date'}
          changeHandler={releaseDateChangeHandler}
          value={releaseDate}
          classes={CLASSES.NETFLIX_APP__INPUT_BLOCK}
        />
        <InputWithLabel
          changeHandler={ratingChangeHandler}
          placeholder={'1123'}
          labelText={'rating'}
          value={rating.toString()}
          classes={CLASSES.NETFLIX_APP__INPUT_BLOCK}
        />
        <InputWithLabel
          changeHandler={runTimeChangeHandler}
          placeholder={'runtime'}
          labelText={'runtime'}
          value={runTime}
          classes={CLASSES.NETFLIX_APP__INPUT_BLOCK}
        />
        <TextAreaWithLabel
          changeHandler={overviewChangeHandler}
          placeholder={'Movie description'}
          labelText={'runtime'}
          value={runTime}
          classes={CLASSES.NETFLIX_APP__INPUT_BLOCK}
        />
      </div>
      <div className={CLASSES.NETFLIX_APP_MODAL_WINDOW_FORM_BUTTONS_SECTION}>
        <Button clickHandler={resetButtonClickHandler} text={'reset'} />
        <Button clickHandler={subminButtonClickHandler} text={'submit'} />
      </div>
    </form>
  );
};

export default Form;