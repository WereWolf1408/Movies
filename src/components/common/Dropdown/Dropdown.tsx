import React, { useEffect, useState } from 'react';
import { DropdownCallbackProps, InputFormProps } from '../../../utils/interfaces';
import { UseFormRegister, Path } from 'react-hook-form';
import classNames from 'classnames';
import { ValidationFormOptionsProps } from '../../../utils/interfaces';

import './Dropdown.less';

const CLASSES = {
  NETFLIX_APP_DROPDOWN: 'netflix-app__dropdown',
  NETFLIX_APP_DROPDOWN_OPTIONS: 'netflix-app__dropdown-options',
  NETFLIX_APP_DROPDOWN_FIELD: 'netflix-app__dropdown-field',
  NETFLIX_APP_DROPDOWN_CHECKBOX: 'netflix-app__dropdown-checkbox',
  NETFLIX_APP_DROPDOWN_CHECKBOX_LABEL: 'netflix-app__dropdown-checkbox-label',
};

interface DropdownProps {
  (props: {
    options: Array<string>;
    dropdownType: 'simple' | 'multiline';
    label?: Path<InputFormProps>;
    validationOptions?: ValidationFormOptionsProps;
    errors?: string;
    selectedValue?: string;
    callback?: DropdownCallbackProps;
    register?: UseFormRegister<InputFormProps>;
  }): JSX.Element;
}

export const Dropdown: DropdownProps = ({
  callback,
  options,
  dropdownType,
  register,
  label,
  validationOptions,
  errors,
  selectedValue,
}) => {
  const [selected, setSelected] = useState<string>('select value');
  const [isOpen, setIsOpen] = useState(false);
  const [optionsHolder, setOptionsHolder] = useState([]);

  useEffect(() => {
    if (selectedValue) {
      setSelected(selectedValue);
    } else {
      setSelected('select value');
    }
  }, [selectedValue]);

  const clickHandler = (option: string) => {
    setSelected(option);
    setIsOpen(false);
    callback && callback(option);
  };

  const multilineClickHandler = (
    option: string,
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    const { checked } = event.currentTarget;
    let newArray = [];
    if (checked) {
      optionsHolder.push(option);
      newArray = [...optionsHolder];
    } else {
      newArray = optionsHolder.filter((value) => value !== option);
    }
    setOptionsHolder(newArray);
    setSelected(newArray.toString());
    callback && callback(newArray);
  };

  const selectedOptionClickHandler = () => {
    setIsOpen((prevVal) => !prevVal);
  };

  const renderDropdownOutputElement = () => {
    if (dropdownType === 'simple') {
      return <span onClick={selectedOptionClickHandler}>{selected}</span>;
    }
    return (
      <>
        <input
          className={'netflix-app__input'}
          readOnly={true}
          {...register(label, validationOptions)}
          onClick={selectedOptionClickHandler}
          value={selected}
          placeholder={'select value'}
        />
        {errors && (
          <span className={'netflix-app__input-error-label'}>
            This is filed is required
          </span>
        )}
      </>
    );
  };

  const simpleDropdownOptionView = (item: string, index: number) => (
    <span
      key={index}
      className={classNames({ ['active']: item === selected })}
      onClick={() => clickHandler(item)}
    >
      {item}
    </span>
  );

  const multilineDropdownOptionView = (item: string, index: number) => (
    <div
      key={index}
      className={classNames({ ['active']: optionsHolder.includes(item) })}
    >
      <label htmlFor="" className={CLASSES.NETFLIX_APP_DROPDOWN_CHECKBOX_LABEL}>
        <input
          className={CLASSES.NETFLIX_APP_DROPDOWN_CHECKBOX}
          type="checkbox"
          value={item}
          checked={optionsHolder.includes(item)}
          onChange={(event) => multilineClickHandler(item, event)}
          role={'checkbox'}
        />
        {item}
      </label>
    </div>
  );

  return (
    <div
      className={classNames(CLASSES.NETFLIX_APP_DROPDOWN, {
        [CLASSES.NETFLIX_APP_DROPDOWN_FIELD]: dropdownType === 'multiline',
      })}
    >
      {renderDropdownOutputElement()}
      {isOpen && (
        <div className={CLASSES.NETFLIX_APP_DROPDOWN_OPTIONS}>
          {options.map((item, index) => {
            return dropdownType === 'simple'
              ? simpleDropdownOptionView(item, index)
              : multilineDropdownOptionView(item, index);
          })}
        </div>
      )}
    </div>
  );
};
