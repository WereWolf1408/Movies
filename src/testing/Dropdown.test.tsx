import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import { store } from '../store/store';
import { Dropdown } from '../components/common/Dropdown';

describe('Dropdown Component', () => {
  it('---> shapshot test:', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Dropdown options={['1', '2', '3']} dropdownType={'simple'} />
      </Provider>
    );
    expect(asFragment).toMatchSnapshot();
  });

  it(`----> test callback trigger correctly:`, () => {
    const mockCallback = jest.fn(() => {
      console.log('simply to test console');
    });
    const { getByText } = render(
      <Provider store={store}>
        <Dropdown
          options={['1', '2', '3']}
          dropdownType={'simple'}
          callback={mockCallback}
        />
      </Provider>
    );
    const dropdown = getByText('select value');
    expect(dropdown).toBeInTheDocument();
    userEvent.click(dropdown);
    const firstOption = getByText('1');
    expect(firstOption).toBeInTheDocument();
    userEvent.click(firstOption);
    expect(mockCallback.mock.calls.length).toBe(1);
  });

  it('test multuselect dropdown:', () => {
    const register = jest.fn();
    const { getByText, getByPlaceholderText, getAllByRole } = render(
      <Provider store={store}>
        <Dropdown
          options={['1', '2', '3']}
          dropdownType={'multiline'}
          register={register}
        />
      </Provider>
    );

    userEvent.click(getByPlaceholderText('select value'));
    expect(getAllByRole('checkbox').length).toBe(3);
    userEvent.click(getAllByRole('checkbox')[0]);
    userEvent.click(getAllByRole('checkbox')[1]);
    expect(getByText('2')).toBeInTheDocument();
    expect(getByText('1')).toBeInTheDocument();
  });
});
