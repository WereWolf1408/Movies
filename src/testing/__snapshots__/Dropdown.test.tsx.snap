// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Component ---> shapshot test: 1`] = `[Function]`;
